# AdvCiv-SAS reachable-history commit diff
# Email privacy: email-shaped addresses redacted from generated commit-history text.
# History metadata layout: 3 (history-segment label; title-only diff header; full message redirected to anonymized Git logs).
# Cache format: 2
# Cache policy: 8c87d6e4bd0f
# Practical commit count: 3509
# Commit: e658c45fb936df8458c34f67e305f2e166fd23e9
# History segment: pre-SAS AdvCiv history
# Parent used for diff: 32857921420d2fcd21120d057501dd245f730a28
# Title: Redistribute AI weight for unavailable victory strats
# Full commit message/metadata: see _0_Common_Docs/git_logs/git_log_anonymized_email_002_Base_AdvCiv.txt.
# History segments describe ancestry/log partition, not authorship: SASBranch includes SAS work plus later upstream AdvCiv commits merged/imported into this branch.
# Practical commit counts can repeat on divergent/merged history; the full Git SHA is the canonical unique commit identifier.
# Diff policy: preserve textual source/docs/config history; summarize known generated/binary/noisy or exceptionally huge file patches.
# Git textconv/external diff drivers are disabled; redundant document formats such as ODT/PDF are summarized rather than converted.

[CHANGE SUMMARY - NUMSTAT + RENAME/MODE SUMMARY]
3	1	CvGameCoreDLL/CvCityAI.cpp
35	0	CvGameCoreDLL/CvGame.cpp
1	0	CvGameCoreDLL/CvGame.h
7	0	CvGameCoreDLL/CvGameAI.cpp
1	0	CvGameCoreDLL/CvGameAI.h
1	14	CvGameCoreDLL/CvGameInterface.cpp
168	85	CvGameCoreDLL/CvPlayerAI.cpp
8	0	CvGameCoreDLL/CvPlayerAI.h
43	14	CvGameCoreDLL/CvUnitAI.cpp

[PATCHES]

diff --git a/CvGameCoreDLL/CvCityAI.cpp b/CvGameCoreDLL/CvCityAI.cpp
index db9d02d1a..46f65dfaa 100644
--- a/CvGameCoreDLL/CvCityAI.cpp
+++ b/CvGameCoreDLL/CvCityAI.cpp
@@ -6360,7 +6360,9 @@ int CvCityAI::AI_projectValue(ProjectTypes eProject) /* advc: */ const
 						if (kLoopPlayer.getID() == kOwner.getID())
 						{
 							iTemp = GC.getInfo(kOwner.getPersonalityType()).
-									// victory weight is between 0 and 100. (usually around 30).
+									/*	victory weight is between 0 and 100. (usually around 30).
+										(advc, note: Just for flavor, don't check whether
+										conquest is actually possible.) */
 									getConquestVictoryWeight() / 2
 									/*  advc.650: Was just 85. More civs =>
 										more targets to choose from. */

diff --git a/CvGameCoreDLL/CvGame.cpp b/CvGameCoreDLL/CvGame.cpp
index abce4cd7b..6056baf3e 100644
--- a/CvGameCoreDLL/CvGame.cpp
+++ b/CvGameCoreDLL/CvGame.cpp
@@ -951,6 +951,7 @@ void CvGame::initFreeState()
 			it->initFreeState();
 	}
 	applyOptionEffects(); // advc.310
+	AI().AI_updateVictoryWeights(); // advc.115f
 	FOR_EACH_ENUM(Tech)
 	{
 		// <advc.126> Later-era free tech only for later-era starts.
@@ -3608,6 +3609,39 @@ bool CvGame::isDiploVictoryValid() const
 	return false;
 }
 
+// advc.115f:
+VictoryTypes CvGame::getDominationVictory() const
+{
+	FOR_EACH_ENUM(Victory)
+	{
+		if (GC.getInfo(eLoopVictory).getLandPercent() > 0 &&
+			GC.getInfo(eLoopVictory).getPopulationPercentLead())
+		{
+			return eLoopVictory;
+		}
+	}
+	return NO_VICTORY;
+}
+
+// advc: Moved from CvGameInterface.cpp
+VictoryTypes CvGame::getSpaceVictory() const
+{
+	FOR_EACH_ENUM(Project)
+	{
+		if (GC.getInfo(eLoopProject).isSpaceship())
+			return GC.getInfo(eLoopProject).getVictoryPrereq();
+	}
+	FErrorMsg("Invalid space victory type");
+	return NO_VICTORY;
+	/*	advc: Alternative check, cut from AI_calculateSpaceVictoryStage (BBAI),
+		disused. */
+	/*FOR_EACH_ENUM(Victory) {
+		if (GC.getInfo(eLoopVictory).getVictoryDelayTurns() > 0)
+			return eLoopVictory;
+	}
+	return NO_VICTORY;*/
+}
+
 
 bool CvGame::isTeamVote(VoteTypes eVote) const
 {
@@ -6179,6 +6213,7 @@ void CvGame::makeCorporationFounded(CorporationTypes eIndex, PlayerTypes ePlayer
 void CvGame::setVictoryValid(VictoryTypes eVict, bool bValid)
 {
 	GC.getInitCore().setVictory(eVict, bValid);
+	AI().AI_updateVictoryWeights(); // advc.115f
 }

diff --git a/CvGameCoreDLL/CvGame.h b/CvGameCoreDLL/CvGame.h
index 55d6822a1..5185dac00 100644
--- a/CvGameCoreDLL/CvGame.h
+++ b/CvGameCoreDLL/CvGame.h
@@ -804,6 +804,7 @@ public:
 	DllExport ColorTypes getPlotHighlightColor(CvPlot* pPlot) const;
 	DllExport void cheatSpaceship() const;
 	DllExport VictoryTypes getSpaceVictory() const;
+	VictoryTypes getDominationVictory() const; // advc.115f
 	DllExport void nextActivePlayer(bool bForward);
 
 	DllExport DomainTypes getUnitDomain(UnitTypes eUnit) const; // advc.003j: Isn't and imo shouldn't be used DLL-internally

diff --git a/CvGameCoreDLL/CvGameAI.cpp b/CvGameCoreDLL/CvGameAI.cpp
index 40590357f..9f106f4f1 100644
--- a/CvGameCoreDLL/CvGameAI.cpp
+++ b/CvGameCoreDLL/CvGameAI.cpp
@@ -224,6 +224,13 @@ scaled CvGameAI::AI_exclusiveRadiusWeight(int iDist) const
 	return m_arExclusiveRadiusWeight[iDist];
 }
 
+// advc.115f:
+void CvGameAI::AI_updateVictoryWeights()
+{
+	for (PlayerAIIter<CIV_ALIVE> it; it.hasNext(); ++it)
+		it->AI_updateVictoryWeights();
+}
+
 // advc.009b:
  void CvGameAI::AI_updateExclusiveRadiusWeight()
  {

diff --git a/CvGameCoreDLL/CvGameAI.h b/CvGameCoreDLL/CvGameAI.h
index eb9682220..480135e25 100644
--- a/CvGameCoreDLL/CvGameAI.h
+++ b/CvGameCoreDLL/CvGameAI.h
@@ -31,6 +31,7 @@ public:
 	EraTypes AI_getVoteSourceEra(VoteSourceTypes eVS = NO_VOTESOURCE) const;
 	// </advc.erai>
 	scaled AI_exclusiveRadiusWeight(int iDist = -1) const; // advc.099b
+	void AI_updateVictoryWeights(); // advc.115f
 
 	void read(FDataStreamBase* pStream);
 	void write(FDataStreamBase* pStream);

diff --git a/CvGameCoreDLL/CvGameInterface.cpp b/CvGameCoreDLL/CvGameInterface.cpp
index 232a4f4b0..372e08d96 100644
--- a/CvGameCoreDLL/CvGameInterface.cpp
+++ b/CvGameCoreDLL/CvGameInterface.cpp
@@ -2344,20 +2344,7 @@ CivilopediaWidgetShowTypes CvGame::getWidgetShow(ImprovementTypes eImprovement)
 	return eType;
 }
 
-VictoryTypes CvGame::getSpaceVictory() const
-{
-	VictoryTypes eVictory = NO_VICTORY;
-	FOR_EACH_ENUM(Project)
-	{
-		if (GC.getInfo(eLoopProject).isSpaceship())
-		{
-			eVictory = GC.getInfo(eLoopProject).getVictoryPrereq();
-			break;
-		}
-	}
-	FAssertMsg(eVictory != NO_VICTORY, "Invalid space victory type");
-	return eVictory;
-}
+// (advc: getSpaceVictory implementation moved to CvGame.cpp)
 
 // advc (note): Alt+Z causes the EXE to call this function
 void CvGame::nextActivePlayer(bool bForward)

diff --git a/CvGameCoreDLL/CvPlayerAI.cpp b/CvGameCoreDLL/CvPlayerAI.cpp
index d0c6b2b22..ae7f5ea77 100644
--- a/CvGameCoreDLL/CvPlayerAI.cpp
+++ b/CvGameCoreDLL/CvPlayerAI.cpp
@@ -299,6 +299,7 @@ void CvPlayerAI::AI_reset(bool bConstructor)
 	{
 		m_aiUnitCombatWeights[iI] = 0;
 	}
+	m_aiVictoryWeights.reset(); // advc.115f
 
 	/*for (iI = 0; iI < MAX_PLAYERS; iI++) {
 		m_aiCloseBordersAttitude[iI] = 0;
@@ -8841,6 +8842,8 @@ PlayerVoteTypes CvPlayerAI::AI_diploVote(const VoteSelectionSubData& kVoteData,
 							(12 +
 							getCapital()->getDefyResolutionAngerTimer() +
 							scaled(kPersonality.getBasePeaceWeight(), 2) +
+							/*	Just for flavor; don't check whether diplo victory
+								is actually enabled or realistic. */
 							scaled(kPersonality.getDiplomacyVictoryWeight(), 10)) *
 							rCityRatio).round();
 					if (AI_isFinancialTrouble())
@@ -14705,7 +14708,7 @@ int CvPlayerAI::AI_nukeWeight() const
 			GC.getGame().getRecommendedPlayers()); // advc.137
 	iHistory = std::min(iHistory, 300);
 	if (iHistory > 0)
-	{
+	{	// advc (note): Just for flavor; don't check whether conquest is possible.
 		iHistory *= 90 + kPersonality.getConquestVictoryWeight();
 		iHistory /= 100;
 	}
@@ -22449,6 +22452,10 @@ void CvPlayerAI::read(FDataStreamBase* pStream)
 
 	pStream->Read(GC.getNumUnitClassInfos(), m_aiUnitClassWeights);
 	pStream->Read(GC.getNumUnitCombatInfos(), m_aiUnitCombatWeights);
+	// <advc.115f>
+	if (uiFlag >= 20)
+		m_aiVictoryWeights.read(pStream);
+	else AI_updateVictoryWeights(); // </advc.115f>
 	// <advc.130n>
 	if (uiFlag == 18)
 	{
@@ -22548,7 +22555,8 @@ void CvPlayerAI::write(FDataStreamBase* pStream)
 	//uiFlag = 16; // advc.651
 	//uiFlag = 17; // advc.550g
 	//uiFlag = 18; // advc.130n
-	uiFlag = 19; // advc.130n (mostly removed again)
+	//uiFlag = 19; // advc.130n (mostly removed again)
+	uiFlag = 20; // advc.115f
 	pStream->Write(uiFlag);
 
 	pStream->Write(m_iPeaceWeight);
@@ -22630,6 +22638,7 @@ void CvPlayerAI::write(FDataStreamBase* pStream)
 
 	pStream->Write(GC.getNumUnitClassInfos(), m_aiUnitClassWeights);
 	pStream->Write(GC.getNumUnitCombatInfos(), m_aiUnitCombatWeights);
+	m_aiVictoryWeights.write(pStream); // advc.115f
 	// K-Mod. save great person weights.
 	{
 		int iItems = m_GreatPersonWeights.size();
@@ -23436,18 +23445,17 @@ int CvPlayerAI::AI_calculateCultureVictoryStage(
 {
 	PROFILE_FUNC();
 
-	CvGame const& kGame = GC.getGame();
-	if (!isHuman() && // advc.115d
-		!kGame.isOption(GAMEOPTION_RISE_FALL) && // advc.700
-		!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_CULTURE"))
-	{
-		return 0;
-	}
-	if (!kGame.culturalVictoryValid())
-		return 0;
-
 	if (!hasCapital())
 		return 0;
+	// <advc.115f>
+	int iWeight = 0;
+	FOR_EACH_ENUM(Victory)
+	{
+		if (GC.getInfo(eLoopVictory).getCityCulture() > 0)
+			iWeight += AI_getVictoryWeight(eLoopVictory);
+	}
+	if (iWeight < 0)
+		return 0; // </advc.115f>
 
 	int iHighCultureCount = 0;
 	int iCloseToLegendaryCount = 0;
@@ -23456,6 +23464,7 @@ int CvPlayerAI::AI_calculateCultureVictoryStage(
 	/*	some (arbitrary) fraction of the game, after which we get more serious.
 		(cf. old code) */
 	int iEraThresholdPercent = 80 - (AI_getStrategyRand(1) % 2) * 20;
+	CvGame const& kGame = GC.getGame();
 	int iLegendaryCulture = kGame.getCultureThreshold(
 			CvCultureLevelInfo::finalCultureLevel());
 	int iVictoryCities = kGame.culturalVictoryNumCultureCities();
@@ -23544,7 +23553,8 @@ int CvPlayerAI::AI_calculateCultureVictoryStage(
 		then we need to consider whether or not to aim for culture at all. */
 	if (iCloseToLegendaryCount < iVictoryCities)
 	{
-		int iValue = GC.getInfo(getPersonalityType()).getCultureVictoryWeight();
+		int iValue = //GC.getInfo(getPersonalityType()).getCultureVictoryWeight()
+				iWeight; // advc.115f
 
 		//if (kGame.isOption(GAMEOPTION_ALWAYS_PEACE))
 		if (!GET_TEAM(getTeam()).AI_isWarPossible()) // advc.001j
@@ -23707,32 +23717,15 @@ int CvPlayerAI::AI_calculateCultureVictoryStage(
 
 int CvPlayerAI::AI_calculateSpaceVictoryStage() const
 {
-	if (!isHuman() && // advc.115d
-		!GC.getGame().isOption(GAMEOPTION_RISE_FALL) && // advc.700
-		!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_SPACE"))
-	{
-		return 0;
-	}
 	if (!hasCapital())
 		return 0;
-
-	// Better way to determine if spaceship victory is valid?
-	VictoryTypes eSpace = NO_VICTORY;
-	FOR_EACH_ENUM(Victory)
-	{
-		if (GC.getGame().isVictoryValid(eLoopVictory))
-		{
-			CvVictoryInfo& kVictoryInfo = GC.getInfo(eLoopVictory);
-			if (kVictoryInfo.getVictoryDelayTurns() > 0)
-			{
-				eSpace = eLoopVictory;
-				break;
-			}
-		}
-	}
-
+	// <advc.115f>
+	VictoryTypes const eSpace = GC.getGame().getSpaceVictory();
 	if (eSpace == NO_VICTORY)
 		return 0;
+	int const iWeight = AI_getVictoryWeight(eSpace);
+	if (iWeight < 0)
+		return 0; // </advc.115f>
 
 	// If have built Apollo, then the race is on!
 	bool bHasApollo = false;
@@ -23827,7 +23820,8 @@ int CvPlayerAI::AI_calculateSpaceVictoryStage() const
 				iSpaceTeams++;
 		}
 		if (200 * iSpaceTeams / (1+iKnownTeams) <=
-			GC.getInfo(getPersonalityType()).getSpaceVictoryWeight() +
+			//GC.getInfo(getPersonalityType()).getSpaceVictoryWeight()
+			iWeight + // advc.115f
 			AI_getStrategyRand(3) % 100 && // note, correlated with number used lower down.
 			(bEnoughProduction || iOurProgress > 10)) // advc.115
 		{
@@ -23889,31 +23883,24 @@ int CvPlayerAI::AI_calculateConquestVictoryStage() const
 {
 	PROFILE_FUNC();
 
-	CvGame const& kGame = GC.getGame(); // advc
+	CvGame const& kGame = GC.getGame();
 	CvTeamAI const& kTeam = GET_TEAM(getTeam());
-
 	// check for validity of conquest victory
 	//if (kGame.isOption(GAMEOPTION_ALWAYS_PEACE)
 	if (!kTeam.AI_isWarPossible() || // advc.001j
-		isAVassal() || (/* advc.115d: */ !isHuman() &&
-		!kGame.isOption(GAMEOPTION_RISE_FALL) && // advc.700
-		!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_CONQUEST")))
+		isAVassal())
 	{
 		return 0;
 	}
+	// <advc.115f>
+	int iWeight = 0;
+	FOR_EACH_ENUM(Victory)
 	{
-		bool bValid = false;
-		FOR_EACH_ENUM(Victory)
-		{
-			if (kGame.isVictoryValid(eLoopVictory) &&
-				GC.getInfo(eLoopVictory).isConquest())
-			{
-				bValid = true;
-			}
-		}
-		if (!bValid)
-			return 0;
+		if (GC.getInfo(eLoopVictory).isConquest())
+			iWeight += AI_getVictoryWeight(eLoopVictory);
 	}
+	if (iWeight < 0)
+		return 0; // </advc.115f>
 
 	/* Things to consider:
 	 - personality (conquest victory weight)
@@ -23998,7 +23985,8 @@ int CvPlayerAI::AI_calculateConquestVictoryStage() const
 	// This decision is based on situation and on personality.
 
 	// personality & randomness
-	int iValue = GC.getInfo(getPersonalityType()).getConquestVictoryWeight();
+	int iValue = //GC.getInfo(getPersonalityType()).getConquestVictoryWeight()
+			iWeight; // advc.115f
 	//iValue += (kGame.isOption(GAMEOPTION_AGGRESSIVE_AI) ? 20 : 0);
 	iValue += (AI_getStrategyRand(4) % 100);
 
@@ -24082,34 +24070,21 @@ int CvPlayerAI::AI_calculateConquestVictoryStage() const
 
 int CvPlayerAI::AI_calculateDominationVictoryStage() const
 {
-	CvGame const& kGame = GC.getGame(); // advc
+	CvGame const& kGame = GC.getGame();
 	CvTeamAI const& kTeam = GET_TEAM(getTeam());
 	//if (kGame.isOption(GAMEOPTION_ALWAYS_PEACE))
 	if (!kTeam.AI_isWarPossible() || // advc.001j
-		isAVassal() || (/* advc.115d: */ !isHuman() &&
-		!kGame.isOption(GAMEOPTION_RISE_FALL) && // advc.700
-		GC.getDefineINT("BBAI_VICTORY_STRATEGY_DOMINATION") <= 0))
+		isAVassal())
 	{
 		return 0;
 	}
-	VictoryTypes eDomination = NO_VICTORY;
-	
-	FOR_EACH_ENUM(Victory)
-	{
-		if (kGame.isVictoryValid(eLoopVictory))
-		{
-			CvVictoryInfo& kVictoryInfo = GC.getInfo(eLoopVictory);
-			if (kVictoryInfo.getLandPercent() > 0 &&
-				kVictoryInfo.getPopulationPercentLead())
-			{
-				eDomination = eLoopVictory;
-				break;
-			}
-		}
-	}
-
+	// <advc.115f>
+	VictoryTypes const eDomination = kGame.getDominationVictory();
 	if (eDomination == NO_VICTORY)
 		return 0;
+	int const iWeight = AI_getVictoryWeight(eDomination);
+	if (iWeight < 0)
+		return 0; // </advc.115f>
 
 	int iPercentOfDomination = 0;
 	int iOurPopPercent = (100 * kTeam.getTotalPopulation()) /
@@ -24154,7 +24129,8 @@ int CvPlayerAI::AI_calculateDominationVictoryStage() const
 
 	// Check for whether we are inclined to pursue a domination strategy
 	{
-		int iValue = GC.getInfo(getPersonalityType()).getDominationVictoryWeight();
+		int iValue = //GC.getInfo(getPersonalityType()).getDominationVictoryWeight()
+				iWeight; // advc.115f
 		iValue += (kGame.isOption(GAMEOPTION_AGGRESSIVE_AI) ?
 				10 : 0); // advc.019: was ?20:0
 		//int iNonsense = AI_getStrategyRand() + 70;
@@ -24177,20 +24153,19 @@ int CvPlayerAI::AI_calculateDominationVictoryStage() const
 
 int CvPlayerAI::AI_calculateDiplomacyVictoryStage() const
 {
-	//int iValue = 0; // advc.115b
-	CvGame const& kGame = GC.getGame(); // advc
-	if (!isHuman() && // advc.115d
-		!kGame.isOption(GAMEOPTION_RISE_FALL) && // advc.700
-		!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_DIPLOMACY"))
+	// <advc.115f>
+	int iWeight = 0;
+	FOR_EACH_ENUM(Victory)
 	{
-		return 0;
+		if (GC.getInfo(eLoopVictory).isDiploVote())
+			iWeight += AI_getVictoryWeight(eLoopVictory);
 	}
+	if (iWeight < 0)
+		return 0; // </advc.115f>
 	// <advc.115c>
 	if (TeamIter<CIV_ALIVE>::count() <= 2)
 		return 0; // </advc.115c>
-	// <advc.178> Code moved into subroutine
-	if (!kGame.isDiploVictoryValid())
-		return false; // </advc.178>
+	CvGame const& kGame = GC.getGame();
 	// Check for whether we are eligible for election
 	bool bVoteEligible = false;
 	FOR_EACH_ENUM(VoteSource)
@@ -24211,11 +24186,15 @@ int CvPlayerAI::AI_calculateDiplomacyVictoryStage() const
 	int iValue = 0;
 	if(isHuman()) // Perhaps not inclined, but very capable.
 		iValue += 100;
-	else iValue += GC.getInfo(getPersonalityType()).getDiplomacyVictoryWeight();
+	else
+	{
+		iValue += //GC.getInfo(getPersonalityType()).getDiplomacyVictoryWeight();
+				iWeight; // advc.115f
+	}
 	iValue = intdiv::round(iValue * 2, 3); // Victory weight too high
 	int iVoteTarget=MIN_INT;
 	int iVotesToGo = GET_TEAM(getTeam()).AI_votesToGoForVictory(&iVoteTarget);
-	if(iVoteTarget == -1)
+	if (iVoteTarget == -1)
 		return 0;
 	FAssert(iVoteTarget > 0);
 	scaled rProgressRatio(iVoteTarget - iVotesToGo, iVoteTarget);
@@ -24318,6 +24297,107 @@ bool CvPlayerAI::AI_atVictoryStage3() const
 			AI_VICTORY_MILITARY3 | AI_VICTORY_DIPLOMACY4);
 }
 
+// advc.115f: In part moved from the AI_calculate...VictoryStage functions
+void CvPlayerAI::AI_updateVictoryWeights()
+{
+	EagerEnumMap<VictoryTypes,bool> abValid;
+	EagerEnumMap<VictoryTypes,short> aiWeight;
+	FOR_EACH_ENUM(Victory)
+	{
+		int iWeight;
+		abValid.set(eLoopVictory, AI_isVictoryValid(eLoopVictory, iWeight));
+		aiWeight.set(eLoopVictory, std::max<short>(0, safeIntCast<short>(iWeight)));
+	}
+	// (not short b/c operator+= doesn't handle short well)
+	int iDeadWeight = 0;
+	int iValidWeight = 0;
+	FOR_EACH_ENUM(Victory)
+	{
+		bool bValid = abValid.get(eLoopVictory);
+		short iWeight = aiWeight.get(eLoopVictory);
+		(bValid ? iValidWeight : iDeadWeight) +=
+				iWeight;
+		m_aiVictoryWeights.set(eLoopVictory,
+				bValid ? iWeight : -1);
+	}
+	if (iValidWeight <= 0 || iDeadWeight <= 0)
+		return;
+	/*	Redistribute the dead weight among the valid victories
+		(the main point of this function). */
+	FOR_EACH_ENUM(Victory)
+	{
+		if (abValid.get(eLoopVictory))
+		{
+			short iWeight = m_aiVictoryWeights.get(eLoopVictory);
+			iWeight = static_cast<short>(iWeight + (iDeadWeight * iWeight) / iValidWeight);
+			m_aiVictoryWeights.set(eLoopVictory, iWeight);
+		}
+	}
+}
+
+// advc.115f:
+bool CvPlayerAI::AI_isVictoryValid(VictoryTypes eVictory, int& iWeight) const
+{
+	iWeight = 0;
+	CvGame const& kGame = GC.getGame();
+	CvLeaderHeadInfo const& kPersonality = GC.getInfo(getPersonalityType());
+	bool const bHuman = isHuman();
+	// Let AI assume that humans assign this weight to every valid victory condition
+	int const iHumanWeight = 30;
+	CvVictoryInfo const& kVictory = GC.getInfo(eVictory);
+	bool const bCheckBBAIDefine = (!bHuman && // advc.115d
+		!kGame.isOption(GAMEOPTION_RISE_FALL)); // advc.700
+	if (kVictory.getCityCulture() > 0)
+	{	/*	(Would be cleaner to replace these 5 getters with a single one
+			that takes a VictoryTypes param.) */
+		iWeight = (bHuman ? iHumanWeight : kPersonality.getCultureVictoryWeight());
+		if (!kGame.culturalVictoryValid())
+			return false;
+		if (bCheckBBAIDefine && 
+			!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_CULTURE"))
+		{
+			return false;
+		}
+	}
+	if (eVictory == kGame.getSpaceVictory())
+	{
+		iWeight = (bHuman ? iHumanWeight : kPersonality.getSpaceVictoryWeight());
+		if (bCheckBBAIDefine &&
+			!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_SPACE"))
+		{
+			return false;
+		}
+	}
+	if (GC.getInfo(eVictory).isConquest())
+	{
+		iWeight = iWeight = (bHuman ? iHumanWeight : kPersonality.getConquestVictoryWeight());
+		if (bCheckBBAIDefine &&
+			!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_CONQUEST"))
+		{
+			return false;
+		}
+	}
+	if (eVictory == kGame.getDominationVictory())
+	{
+		iWeight = iWeight = (bHuman ? iHumanWeight : kPersonality.getDominationVictoryWeight());
+		if (bCheckBBAIDefine &&
+			!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_DOMINATION"))
+		{
+			return false;
+		}
+	}
+	if (kVictory.isDiploVote())
+	{
+		iWeight = iWeight = (bHuman ? iHumanWeight : kPersonality.getDiplomacyVictoryWeight());
+		if (bCheckBBAIDefine &&
+			!GC.getDefineBOOL("BBAI_VICTORY_STRATEGY_DIPLOMACY"))
+		{
+			return false;
+		}
+	}
+	return kGame.isVictoryValid(eVictory);
+}
+
 
 void CvPlayerAI::AI_updateVictoryStageHash()
 {
@@ -28840,7 +28920,10 @@ void CvPlayerAI::AI_setHuman(bool b)
 {
 	// Some of the first-impression modifiers don't apply to human players
 	for (PlayerAIIter<MAJOR_CIV> it; it.hasNext(); ++it)
+	{
 		it->AI_updateAttitude();
+		it->AI_updateVictoryWeights(); // advc.115b, advc.115f
+	}
 	if (b)
 		return;
 	/*	<advc.057> Enforce invariant: AI group head has maximal impassable count.

diff --git a/CvGameCoreDLL/CvPlayerAI.h b/CvGameCoreDLL/CvPlayerAI.h
index c642eebd3..119dd941e 100644
--- a/CvGameCoreDLL/CvPlayerAI.h
+++ b/CvGameCoreDLL/CvPlayerAI.h
@@ -566,6 +566,12 @@ public:
 	bool AI_atVictoryStage4() const;
 	bool AI_atVictoryStage3() const;
 	AIVictoryStage AI_getVictoryStageHash() const { return m_eVictoryStageHash; }
+	// advc.115f:
+	int AI_getVictoryWeight(VictoryTypes eVictory) const
+	{
+		return m_aiVictoryWeights.get(eVictory);
+	}
+	void AI_updateVictoryWeights(); // advc.115f
 	void AI_updateVictoryStageHash(); // K-Mod
 	void AI_initStrategyRand(); // K-Mod
 	int AI_getStrategyRand(int iShift) const;
@@ -791,6 +797,7 @@ protected:
 	int* m_aiBonusValueTrade; // advc.036
 	int* m_aiUnitClassWeights;
 	int* m_aiUnitCombatWeights;
+	ArrayEnumMap<VictoryTypes,short> m_aiVictoryWeights; // advc.115f
 	// <advc.130c>
 	bool m_abTheyFarAhead[MAX_CIV_PLAYERS];
 	bool m_abTheyBarelyAhead[MAX_CIV_PLAYERS]; // </advc.130c>
@@ -926,6 +933,7 @@ protected:
 	int AI_calculateDominationVictoryStage() const;
 	int AI_calculateDiplomacyVictoryStage() const;
 	// BETTER_BTS_AI_MOD: END
+	bool AI_isVictoryValid(VictoryTypes eVictory, int& iWeight) const; // advc.115f
 	// K-Mod. I've moved the bulk of AI_getStrategyHash into a new function: AI_updateStrategyHash.
 	AIStrategy AI_getStrategyHash() const { return m_eStrategyHash; }
 	void AI_updateStrategyHash();

diff --git a/CvGameCoreDLL/CvUnitAI.cpp b/CvGameCoreDLL/CvUnitAI.cpp
index 689fb95ee..1ba5bba59 100644
--- a/CvGameCoreDLL/CvUnitAI.cpp
+++ b/CvGameCoreDLL/CvUnitAI.cpp
@@ -5059,26 +5059,55 @@ void CvUnitAI::AI_greatPersonMove()
 		missions.push_back(std::pair<int, int>(rDiscoverValue.round(), GP_DISCOVER));
 	}
 
-	// SlowValue is meant to be a rough estimation of how much value we'll get from doing the best join / build mission.
-	// To give this estimate, I'm going to do a rough personality-based calculation of how many turns to count.
-	// Note that "iBestValue" is roughly 100x commerce per turn for our best join or build mission.
-	// Also note that the commerce per turn is likely to increase as we improve our city infrastructure and so on.
+	CvGame const& kGame = GC.getGame();
+	/*	SlowValue is meant to be a rough estimation of how much value we'll get
+		from doing the best join / build mission. To give this estimate,
+		I'm going to do a rough personality-based calculation of how many turns
+		to count. Note that "iBestValue" is roughly 100x commerce per turn for
+		our best join or build mission. Also note that the commerce per turn is
+		likely to increase as we improve our city infrastructure and so on. */
 	int iSlowValue = iBestValue;
 	if (iSlowValue > 0)
 	{
 		// multiply by the full number of turns remaining
-		iSlowValue *= GC.getGame().getEstimateEndTurn() - GC.getGame().getGameTurn();
+		iSlowValue *= kGame.getEstimateEndTurn() - kGame.getGameTurn();
 
 		// construct a modifier based on what victory we might like to aim for with our personality & situation
-		const CvLeaderHeadInfo& kLeader = GC.getInfo(kPlayer.getPersonalityType());
-		int iModifier =
-			2 * std::max(kLeader.getSpaceVictoryWeight(), kPlayer.AI_atVictoryStage(AI_VICTORY_SPACE1) ? 35 : 0) +
-			1 * std::max(kLeader.getCultureVictoryWeight(), kPlayer.AI_atVictoryStage(AI_VICTORY_CULTURE1) ? 35 : 0) +
-			//0 * kLeader.getDiplomacyVictoryWeight() +
-			-1 * std::max(kLeader.getDominationVictoryWeight(), kPlayer.AI_atVictoryStage(AI_VICTORY_DOMINATION1) ? 35 : 0) +
-			-2 * std::max(kLeader.getConquestVictoryWeight(), kPlayer.AI_atVictoryStage(AI_VICTORY_CONQUEST1) ? 35 : 0);
-		// If we're small, then slow & steady progress might be our best hope to keep up. So increase the modifier for small civs. (think avg. cities / our cities)
-		iModifier += range(40 * GC.getGame().getNumCivCities() / std::max(1, GC.getGame().countCivPlayersAlive()*kPlayer.getNumCities()) - 50, 0, 50);
+		int iModifier = 0;
+		/*	<advc.115f> Had used weights from kOwner's personality directly.
+			I don't think the weights should count when the respective
+			victory is unavailable (debatable ...). */
+		FOR_EACH_ENUM(Victory)
+		{
+			int const iWeight = kOwner.AI_getVictoryWeight(eLoopVictory);
+			if (eLoopVictory == kGame.getSpaceVictory())
+			{
+				iModifier += 2 * std::max(iWeight,
+						kOwner.AI_atVictoryStage(AI_VICTORY_SPACE1) ? 35 : 0);
+			}
+			if (GC.getInfo(eLoopVictory).getCityCulture() > 0)
+			{
+				iModifier += 1 * std::max(iWeight,
+						kOwner.AI_atVictoryStage(AI_VICTORY_CULTURE1) ? 35 : 0);
+			}
+			if (eLoopVictory == kGame.getDominationVictory())
+			{
+				iModifier -= std::max(iWeight,
+						kOwner.AI_atVictoryStage(AI_VICTORY_DOMINATION1) ? 35 : 0);
+			}
+			if (GC.getInfo(eLoopVictory).isConquest())
+			{
+				iModifier -= 2 * std::max(iWeight,
+						kOwner.AI_atVictoryStage(AI_VICTORY_CONQUEST1) ? 35 : 0);
+			}
+			//if (GC.getInfo(eLoopVictory).isDiploVote()) {}
+		} // </advc.115f>
+		/*	If we're small, then slow & steady progress might be our best hope
+			to keep up. So increase the modifier for small civs.
+			(think avg. cities / our cities) */
+		iModifier += range(40 * kGame.getNumCivCities() / std::max(1,
+				kGame.countCivPlayersAlive() * kOwner.getNumCities()) - 50,
+				0, 50);
 
 		// convert the modifier into some percentage of the remaining turns
 		iModifier = range(30 + iModifier/2, 20, 80);

[PATCH COVERAGE]
patchFiles=9 embeddedFiles=9 omittedFiles=0 embeddedPatchBytes=24149
